File: frontend/src/decisions/types/decisions.ts
// src/types/decisions.ts
import { ImpactLevel } from '../../common/types/common';

export type DecisionType = 'quality' | 'pipeline' | 'security';
export type DecisionStatus = 'pending' | 'completed' | 'deferred' | 'expired';
export type DecisionUrgency = ImpactLevel;
export type DecisionVote = 'approve' | 'reject' | 'defer';

export interface DecisionOption {
  id: string;
  title: string;
  description: string;
  impact: ImpactLevel;
  consequences: string[];
  requirements?: string[];
  estimatedEffort?: string;
  automaticApplicable: boolean;
}

export interface Decision {
  id: string;
  type: DecisionType;
  title: string;
  description: string;
  urgency: DecisionUrgency;
  status: DecisionStatus;
  options: DecisionOption[];
  context: Record<string, unknown>;
  deadline?: string;
  createdAt: string;
  updatedAt: string;
  createdBy: string;
  assignedTo?: string[];
  selectedOption?: string;
  pipelineId: string;
}

export interface DecisionDetails extends Decision {
  analysis?: {
    risks: string[];
    benefits: string[];
    alternatives: string[];
    dependencies?: string[];
  };
  history: DecisionHistoryEntry[];
  votes?: DecisionVote[];
  comments?: DecisionComment[];
}

export interface DecisionHistoryEntry {
  id: string;
  decisionId: string;
  action: 'create' | 'update' | 'vote' | 'comment' | 'apply';
  user: string;
  timestamp: string;
  details: Record<string, unknown>;
  changes?: {
    field: string;
    oldValue: unknown;
    newValue: unknown;
  }[];
}

export interface DecisionComment {
  id: string;
  decisionId: string;
  user: string;
  content: string;
  timestamp: string;
  replyTo?: string;
}

export interface DecisionImpactAnalysis {
  optionId: string;
  risks: Array<{
    description: string;
    severity: ImpactLevel;
    probability: number;
    mitigation?: string;
  }>;
  benefits: Array<{
    description: string;
    impact: ImpactLevel;
    confidence: number;
  }>;
  metrics: Record<string, number>;
  recommendations: string[];
}

export interface DecisionFilters {
  types?: DecisionType[];
  status?: DecisionStatus[];
  urgency?: DecisionUrgency[];
  assignedTo?: string[];
  dateRange?: {
    start: string;
    end: string;
  };
}

export interface DateRange {
  start?: string;
  end?: string;
}

export interface DecisionFilterOptions {
  types?: DecisionType[];
  status?: DecisionStatus[];
  urgency?: DecisionUrgency[];
  assignedTo?: string[];
  dateRange?: DateRange;
}

export interface DecisionState {
  decisions: Record<string, {
    id: string;
    title: string;
    type: string;
    status: 'pending' | 'approved' | 'rejected' | 'implemented';
    urgency: 'high' | 'medium' | 'low';
    description: string;
    assignedTo: string[];
    metadata: {
      createdAt: string;
      updatedAt: string;
      implementedAt?: string;
      rationale: string;
      stakeholders: string[];
    };
    alternatives: Array<{
      id: string;
      description: string;
      pros: string[];
      cons: string[];
    }>;
    impact: {
      areas: Array<{
        name: string;
        impact: 'high' | 'medium' | 'low';
        description: string;
      }>;
      risks: Array<{
        description: string;
        likelihood: 'high' | 'medium' | 'low';
        mitigation: string;
      }>;
    };
    documents: Array<{
      id: string;
      name: string;
      url: string;
    }>;
  }>;
  history: Array<{
    id: string;
    decisionId: string;
    type: 'creation' | 'update' | 'status_change' | 'assignment';
    changes: Record<string, unknown>;
    timestamp: string;
    userId: string;
  }>;
  filters: {
    types?: string[];
    status?: string[];
    urgency?: string[];
    assignedTo?: string[];
    dateRange?: {
      start: string;
      end: string;
    };
  };
  selectedDecisionId: string | null;
  isLoading: boolean;
  error: string | null;
}

File: backend/backend/flask_api/app/schemas/decisions.py
# schemas/decisions.py
from marshmallow import Schema, fields, validate
from .base import BaseRequestSchema, BaseResponseSchema

class DecisionRequestSchema(BaseRequestSchema):
   """Schema for decision request."""
   pipeline_id = fields.UUID(required=True)
   type = fields.String(required=True)
   priority = fields.String(validate=validate.OneOf(['low', 'medium', 'high', 'critical']))
   deadline = fields.DateTime(required=False)
   context = fields.Dict(required=True)

class DecisionResponseSchema(BaseResponseSchema):
   """Schema for decision response."""
   id = fields.UUID(dump_only=True)
   pipeline_id = fields.UUID()
   type = fields.String()
   status = fields.String(validate=validate.OneOf(['pending', 'approved', 'rejected', 'deferred']))
   priority = fields.String()
   deadline = fields.DateTime()
   meta_info = fields.Dict()
   impact_analysis = fields.Dict()
   created_at = fields.DateTime(dump_only=True)
   updated_at = fields.DateTime(dump_only=True)

class DecisionListResponseSchema(BaseResponseSchema):
   """Schema for list of decisions response."""
   decisions = fields.List(fields.Nested(DecisionResponseSchema))
   total_count = fields.Integer()
   pending_count = fields.Integer()

class DecisionHistoryResponseSchema(BaseResponseSchema):
   """Schema for decision history response."""
   history = fields.List(fields.Dict(
       action=fields.String(),
       previous_status=fields.String(),
       new_status=fields.String(),
       user_id=fields.UUID(),
       timestamp=fields.DateTime(),
       meta_data=fields.Dict()
   ))

class DecisionImpactResponseSchema(BaseResponseSchema):
   """Schema for decision impact analysis response."""
   impact = fields.Dict(
       severity=fields.String(validate=validate.OneOf(['low', 'medium', 'high', 'critical'])),
       affected_components=fields.List(fields.String()),
       risk_assessment=fields.Dict(),
       recommendations=fields.List(fields.Dict()),
       metrics=fields.Dict()
   )

class DecisionFeedbackRequestSchema(BaseRequestSchema):
   """Schema for decision feedback request."""
   rating = fields.Integer(validate=validate.Range(min=1, max=5))
   comment = fields.String()
   metrics = fields.Dict()
   suggestions = fields.List(fields.String())

class DecisionFeedbackResponseSchema(BaseResponseSchema):
   """Schema for decision feedback response."""
   id = fields.UUID(dump_only=True)
   decision_id = fields.UUID()
   user_id = fields.UUID()
   rating = fields.Integer()
   comment = fields.String()
   metrics = fields.Dict()
   suggestions = fields.List(fields.String())
   created_at = fields.DateTime(dump_only=True)

class DecisionOptionSchema(BaseRequestSchema):
   """Schema for decision options."""
   name = fields.String(required=True)
   description = fields.String()
   impact_score = fields.Float()
   risks = fields.Dict()
   benefits = fields.Dict()
   is_selected = fields.Boolean(default=False)

class DecisionCommentSchema(BaseRequestSchema):
   """Schema for decision comments."""
   content = fields.String(required=True)
   parent_id = fields.UUID()
   attachments = fields.List(fields.Dict())